### PR TITLE
Changing name in Lobby, finalize ScanQRCode

### PR DESCRIPTION
**Pull request by mkorob**
One note: when updating the name, as there is no update button, it reloads the page every time there is any change to the username (even midtyping)- it´s a bit annoying but I don´t find a better alternative... It could be either OnEnter (which apparently is not great because it´s an unncessary listener), adding a button to submit (possible, but needs to go somewhere), or onOutofFocus like if the mouse clicks out of the form, but that´s not very intuitive I think?

### CODE REVIEW

File: src/components/views/lobby/Lobby.js [hunk 1]
```diff
@@ -61,6 +75,52 @@ const Lobby = ({id}) => {
         }
     }
 
+ 
+
+    // set new name, let the user know if it's already taken (409)
+    const setNewName = async(user, un) => {
```
Comment (apple00juice) [inline src/components/views/lobby/Lobby.js:81]: better name for  un (newName?)

File: src/components/views/lobby/Lobby.js [hunk 2]
```diff
@@ -216,6 +276,13 @@ const Lobby = ({id}) => {
                         Close
                     </Button>
                 </CustomPopUp>
+                <CustomPopUp open={forbiddenChange !== false} information={"Changing information after setting ready status is not possible!"}>
```
Comment (apple00juice) [inline src/components/views/lobby/Lobby.js:279]: Why not use the normal errorMessage popup?, looks like it also just displays a message

**Review by apple00juice — APPROVED** [review]
Rest looks good 👍

**Comment by apple00juice** [comment]
> One note: when updating the name, as there is no update button, it reloads the page every time there is any change to the username (even midtyping)- it´s a bit annoying but I don´t find a better alternative... It could be either OnEnter (which apparently is not great because it´s an unncessary listener), adding a button to submit (possible, but needs to go somewhere), or onOutofFocus like if the mouse clicks out of the form, but that´s not very intuitive I think?

hmmm, that is indeed a difficult...
-> I guess this works, but I haven't experienced how annoying it is
-> onOutofFocus + if the user presses enter would also be an option
-> another option would be that if the user presses the name a popup with a textfield is displayed and there are two buttons save and cancel

**Comment by luis-tm** [comment]
> > One note: when updating the name, as there is no update button, it reloads the page every time there is any change to the username (even midtyping)- it´s a bit annoying but I don´t find a better alternative... It could be either OnEnter (which apparently is not great because it´s an unncessary listener), adding a button to submit (possible, but needs to go somewhere), or onOutofFocus like if the mouse clicks out of the form, but that´s not very intuitive I think?
> 
> hmmm, that is indeed a difficult... -> I guess this works, but I haven't experienced how annoying it is -> onOutofFocus + if the user presses enter would also be an option -> another option would be that if the user presses the name a popup with a textfield is displayed and there are two buttons save and cancel

You're right Masha, it's very annoying how it works now XD

Another proposal: adding a small button next to the name, and differentiate between onClick (button) and onChange (form). onChange will modify a new state (playerName), and only clicking the button will call the server.

const [playerName, setPlayerName] = useState(null);  --> new state

// form for changing name
const FormName = props => {
    return (
        <div className="buttonIn">
            <input
                className="buttonIn name"
                value={props.value}
                onChange={e => props.onChange(e.target.value)}
            />
            <button
                className="buttonIn button"
                onClick={e => props.onClick(e.target.value)}>
            </button>
        </div>
    );
  };

const changeName = async(user) => {
    ...
      const requestBody = { "name": playerName }; --> here we use the state playerName
    ...
}

<FormName
        value={playerName === null ? user.name : playerName} --> we only show the state if it has a value
        onChange={newName => setPlayerName(newName)}
        onClick={() => changeName(user)}
        >
</FormName>

And the css for putting both together:

.buttonIn {
    position: relative;
    height: 32px;

    &.button {
        position: absolute;
        background: $accent;
        right: 0px;
        width: 20px;
    }

    &.name {
        color: $accent;
        text-align: center;
        background: black;
        border: 0.5px solid #FBC12E;
        border-radius: 7px;
        width: 100%;
    }
}

But I also like the idea of the popup 👍

I've put the code in a new branch changename_2 (sorry for such a long comment).

image

File: src/components/views/lobby/Lobby.js (hunk 2 above)
Comment (mkorob) [reply src/components/views/lobby/Lobby.js]: ya that's a good point! It seemed like a good idea to me at the time to add a different message for something that isn't a server-related error, but I guess it's a bit tedious to have it all to be used one time!

**Review by luis-tm — APPROVED** [review]

**Pull request merged**